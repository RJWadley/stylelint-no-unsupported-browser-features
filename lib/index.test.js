/* eslint-env jest */

var stylelint = require('stylelint');
var deleteKeys = require('key-del');

function getOptions(code, rules) {
  return {
    code: code,
    configBasedir: __dirname,
    config: {
      plugins: ['./index.js'],
      rules: rules
    }
  };
}

function parseResult(result) {
  var parsedResult = JSON.parse(result);

  if (parsedResult.length > 1) {
    throw Error('More than one result received');
  }

  return deleteKeys(parsedResult[0], ['source']);
}

describe('stylelint-no-unsupported-browser-features', function() {
  describe('single browser', function() {
    it('should allow display:table for IE 8', function(done) {
      var code = 'div { display: table; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 8']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(false);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });

    it('should disallow rems for IE 9', function(done) {
      var code = 'div { width: 10rem; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 9']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(true);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });

    it('should disallow display:table for IE 6', function(done) {
      var code = 'div { display: table; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 6']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(true);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });
  });

  describe('multiple browsers', function() {
    it('should allow display:table for IE 8 and IE 9', function(done) {
      var code = 'div { display: table; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 8', 'IE 9']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(false);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });

    it('should disallow display:table for IE 6 and IE 7', function(done) {
      var code = 'div { display: table; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 6', 'IE 7']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(true);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });
  });

  describe('ignore option', function() {
    it('should ignore a single property', function(done) {
      var code = 'div { display: flex; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 6'],
            ignore: ['flexbox']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(false);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });

    it('should ignore multiple properties', function(done) {
      var code = 'div { display: flex; } a { display: table; }';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: ['IE 6'],
            ignore: ['flexbox', 'css-table']
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(false);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });
  });

  describe('validate options', function() {
    it('should validate the browsers option', function(done) {
      var code = '';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            browsers: [1]
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(true);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });

    it('should validate the ignore option', function(done) {
      var code = '';
      var rules = {
        'plugin/no-unsupported-browser-features': [
          true,
          {
            ignore: [1]
          }
        ]
      };
      var options = getOptions(code, rules);

      return stylelint
        .lint(options)
        .then(function(result) {
          expect(result.errored).toBe(true);
          expect(parseResult(result.output)).toMatchSnapshot();
          return done();
        })
        .catch(function(error) {
          return done(error);
        });
    });
  });
});
