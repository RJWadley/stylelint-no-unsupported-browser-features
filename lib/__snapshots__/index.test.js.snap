// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stylelint-no-unsupported-browser-features ignore option should ignore a single property 1`] = `
Object {
  "deprecations": Array [],
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [],
}
`;

exports[`stylelint-no-unsupported-browser-features ignore option should ignore multiple properties 1`] = `
Object {
  "deprecations": Array [],
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [],
}
`;

exports[`stylelint-no-unsupported-browser-features multiple browsers should allow display:table for IE 8 and IE 9 1`] = `
Object {
  "deprecations": Array [],
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [],
}
`;

exports[`stylelint-no-unsupported-browser-features multiple browsers should disallow display:table for IE 6 and IE 7 1`] = `
Object {
  "deprecations": Array [],
  "errored": true,
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [
    Object {
      "column": 7,
      "line": 1,
      "rule": "plugin/no-unsupported-browser-features",
      "severity": "error",
      "text": "Unexpected browser feature \\"css-table\\" is not supported by IE 6,7 (plugin/no-unsupported-browser-features)",
    },
  ],
}
`;

exports[`stylelint-no-unsupported-browser-features single browser should allow display:table for IE 8 1`] = `
Object {
  "deprecations": Array [],
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [],
}
`;

exports[`stylelint-no-unsupported-browser-features single browser should disallow display:table for IE 6 1`] = `
Object {
  "deprecations": Array [],
  "errored": true,
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [
    Object {
      "column": 7,
      "line": 1,
      "rule": "plugin/no-unsupported-browser-features",
      "severity": "error",
      "text": "Unexpected browser feature \\"css-table\\" is not supported by IE 6 (plugin/no-unsupported-browser-features)",
    },
  ],
}
`;

exports[`stylelint-no-unsupported-browser-features single browser should disallow rems for IE 9 1`] = `
Object {
  "deprecations": Array [],
  "errored": true,
  "invalidOptionWarnings": Array [],
  "parseErrors": Array [],
  "warnings": Array [
    Object {
      "column": 7,
      "line": 1,
      "rule": "plugin/no-unsupported-browser-features",
      "severity": "error",
      "text": "Unexpected browser feature \\"rem\\" is only partially supported by IE 9 (plugin/no-unsupported-browser-features)",
    },
  ],
}
`;

exports[`stylelint-no-unsupported-browser-features validate options should validate the browsers option 1`] = `
Object {
  "deprecations": Array [],
  "errored": true,
  "invalidOptionWarnings": Array [
    Object {
      "text": "Invalid value \\"1\\" for option \\"browsers\\" of rule \\"plugin/no-unsupported-browser-features\\"",
    },
  ],
  "parseErrors": Array [],
  "warnings": Array [],
}
`;

exports[`stylelint-no-unsupported-browser-features validate options should validate the ignore option 1`] = `
Object {
  "deprecations": Array [],
  "errored": true,
  "invalidOptionWarnings": Array [
    Object {
      "text": "Invalid value \\"1\\" for option \\"ignore\\" of rule \\"plugin/no-unsupported-browser-features\\"",
    },
  ],
  "parseErrors": Array [],
  "warnings": Array [],
}
`;
